src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -497,34 +497,75 @@ int encrypt(byte[] in, int inOfs, int len) {
      * @param len the length of the input data
      * @param out the buffer for the result
      * @param outOfs the offset in <code>out</code>
-     * @exception ProviderException if <code>len</code> is not
-     * a multiple of the block size
      * @return the number of bytes placed into the <code>out</code> buffer
      */
-    int encrypt(byte[] in, int inOfs, int len, byte[] out, int outOfs) {
-        checkDataLength(processed, len);
+    int encrypt(byte[] in, int inOfs, int inLen, byte[] out, int outOfs) {
+        checkDataLength(processed, inLen);
 
         processAAD();
-        int remainder = len % blockSize;
-        if ( remainder > 0) {
+        // 'len' stores the length to use with buffer 'in'.
+        // 'inLen' stores the length returned by the method.
+        int len = inLen;
+
+        // if there is enough data in the ibuffer and 'in' construct a block
+        // to encrypt
+
+        if (ibuffer != null && ibuffer.size() > 0 &&
+            (inLen + ibuffer.size() >= blockSize)) {
+            ArrayUtil.nullAndBoundsCheck(in, inOfs, inLen);
+            ArrayUtil.nullAndBoundsCheck(out, outOfs, inLen);
+            byte[] block = new byte[blockSize];
+            int inLenUsed = constructBlock(ibuffer.toByteArray(), in, block);
+            gctrPAndC.update(block, 0, blockSize, out, outOfs);
+            processed += blockSize;
+            ghashAllToS.update(out, outOfs, blockSize);
+            inOfs += inLenUsed;
+            inLen -= inLenUsed;
+            len += (blockSize - inLenUsed);
+            outOfs += blockSize;
+            ibuffer.reset();
+        }
+
+        // if there it not enough data in the ibuffer and 'in', then append
+        // that to the ibuffer.
+        int remainder = inLen % blockSize;
+        if (remainder > 0) {
             if (ibuffer == null) {
-                ibuffer = new ByteArrayOutputStream(len % blockSize);
+                ibuffer = new ByteArrayOutputStream(inLen % blockSize);
             }
             len -= remainder;
-            ibuffer.write(in, len, remainder);
+            inLen -= remainder;
+            // remainder offset is based on original buffer length
+            ibuffer.write(in, inOfs + inLen, remainder);
         }
```

Comment:
I wonder if this can be moved down for better readability, i.e. process data in multiple of blocks, and store the remaining into 'ibuffer'?

Reply:
I tried to, but I don't like how the variable line up doing the remainder afterwards.  I put some hopefully better comments above each section

Reply:
Ok.